### PR TITLE
feat: implement naive fuzz test for region migration

### PR DESCRIPTION
**Pull request by WenyXu**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## What's changed and what's your intention?

1. Adapt create table generator for new partition rules
2. Implement naive fuzz test for region migration

## Checklist

- [ ] I have written the necessary rustdoc comments.
- [ ] I have added the necessary unit tests and integration tests.
- [ ] This PR requires documentation updates.

## Summary by CodeRabbit

- **New Features**
	- Introduced additional targets for workflow configuration to enhance testing capabilities.
	- Added new functionalities for testing database region migration and failover through fuzz testing.
  
- **Refactor**
	- Improved the logic for generating partition definitions and expressions for tables.
	- Updated partitioning logic and table options formatting for better performance.

- **Tests**
	- Enhanced testing procedures by adding methods for timestamp retrieval, column information extraction, and procedure state management.
	- Improved region migration testing with new asynchronous functions and error handling capabilities.

### CODE REVIEW

**Comment by WenyXu** [comment]
@coderabbitai resolve

File: tests-fuzz/src/ir.rs [hunk 1]
```diff
@@ -496,6 +536,31 @@ pub fn replace_default(
...
+        let b_keys: Vec<_> = primary_keys_idx.iter().map(|&i| &b[i]).collect();
+        for (a_key, b_key) in a_keys.iter().zip(b_keys.iter()) {
+            match a_key.cmp(b_key) {
+                Some(std::cmp::Ordering::Equal) => continue,
+                non_eq => return non_eq.unwrap(),
+            }
+        }
+        std::cmp::Ordering::Equal
+    });
+}
```
Comment (CookiePieWw) [inline tests-fuzz/src/ir.rs:552]: There're similar code snippets in insert fuzz tests. Are they serving the same purpose? Maybe I could refactor the insert fuzz tests with this later.

Comment (WenyXu) [reply tests-fuzz/src/ir.rs:552]: Yes, it's fine for now. we can refactor it in a separate PR later.☺️

File: tests-fuzz/src/utils/migration.rs [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+use crate::utils::partition::{fetch_partitions, region_distribution};
+use crate::utils::wait::wait_condition_fn;
+
+/// Migrates a region from one peer to another within a specified timeout.
+///
+/// Returns the procedure id.
+pub async fn migrate_region(
+    e: &Pool<MySql>,
+    region_id: u64,
+    from_peer: u64,
```
Comment (fengjiachun) [inline tests-fuzz/src/utils/migration.rs:34]:
```suggestion
    from_peer_id: u64,
```

File: tests-fuzz/src/utils/migration.rs [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    to_peer: u64,
```
Comment (fengjiachun) [inline tests-fuzz/src/utils/migration.rs:35]:
```suggestion
    to_peer_id: u64,
```

**Review by fengjiachun — APPROVED** [review]

File: tests-fuzz/src/ir/insert_expr.rs [hunk 4]
```diff
@@ -25,6 +25,41 @@ pub struct InsertIntoExpr {
...
+        self.columns.iter().find(|c| c.is_time_index()).cloned()
+    }
+
+    /// Returns index of the timestamp column
+    pub fn timestamp_column_idx(&self) -> Option<usize> {
+        self.columns
+            .iter()
+            .enumerate()
+            .find(|(_, c)| c.is_time_index())
+            .map(|(idx, _)| idx)
```
Comment (MichaelScofield) [inline tests-fuzz/src/ir/insert_expr.rs:40]:
```suggestion
            .find_map(|(idx, c)| if c.is_time_index() { Some(idx) } else { None })
```

File: tests-fuzz/src/ir/insert_expr.rs [hunk 5]
```diff
@@ -25,6 +25,41 @@ pub struct InsertIntoExpr {
...
+            .collect::<Vec<_>>()
+    }
+
+    /// Returns the indices of columns that are primary keys or time indices.
+    pub fn primary_key_column_idx(&self) -> Vec<usize> {
+        self.columns
+            .iter()
+            .enumerate()
+            .filter(|(_, c)| c.is_primary_key() || c.is_time_index())
+            .map(|(i, _)| i)
```
Comment (MichaelScofield) [inline tests-fuzz/src/ir/insert_expr.rs:58]:
```suggestion
            .filter_map(|(i, c)| if c.is_primary_key() || c.is_time_index() { Some(i) } else { None })
```

File: tests-fuzz/src/ir.rs [hunk 6]
```diff
@@ -496,6 +536,31 @@ pub fn replace_default(
     new_rows
 }
 
+/// Sorts a vector of rows based on the values in the specified primary key columns.
+pub fn sort_by_primary_keys(rows: &mut [RowValues], primary_keys_idx: Vec<usize>) {
+    rows.sort_by(|a, b| {
+        let a_keys: Vec<_> = primary_keys_idx.iter().map(|&i| &a[i]).collect();
+        let b_keys: Vec<_> = primary_keys_idx.iter().map(|&i| &b[i]).collect();
+        for (a_key, b_key) in a_keys.iter().zip(b_keys.iter()) {
```
Comment (evenyag) [inline tests-fuzz/src/ir.rs:544]: `a_key.cmp(&b_key)`?

File: tests-fuzz/src/utils/migration.rs [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+                match region_distribution.get(datanode) {
+                    Some(regions) => {
+                        if expected_regions != regions {
+                            return false;
+                        }
+                    }
+                    None => return false,
+                }
+            }
+            true
```
Comment (evenyag) [inline tests-fuzz/src/utils/migration.rs:87]: Can we just compare `expected_region_distribution` with `region_distribution`?

**Review by MichaelScofield — APPROVED** [review]

File: tests-fuzz/targets/migration/fuzz_migrate_mito_regions.rs [hunk 8]
```diff
@@ -0,0 +1,305 @@
...
+        &ctx.greptime,
+        Duration::from_secs(120),
+        table_ctx.name.clone(),
+        new_distribution,
+    )
+    .await;
+
+    for procedure_id in procedure_ids {
+        wait_for_procedure_finish(&ctx.greptime, Duration::from_secs(120), procedure_id).await;
+    }
```
Comment (evenyag) [inline tests-fuzz/targets/migration/fuzz_migrate_mito_regions.rs:236]: Should we wait procedure before waiting region distribution?

Comment (WenyXu) [reply tests-fuzz/targets/migration/fuzz_migrate_mito_regions.rs:236]: it's okay

**Review by zhongzc — APPROVED** [review]

**Pull request merged**